File: 03-java-socket/src/test/java/HelloProtelisTests.java
import com.uchuhimo.konf.BaseConfig;
import com.uchuhimo.konf.Config;
import demo.*;
import demo.data.ProtelisNode;
import org.apache.commons.math3.util.Pair;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.protelis.lang.ProtelisLoader;
import org.protelis.vm.ProtelisProgram;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;

import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.times;

@ExtendWith(MockitoExtension.class)
class HelloProtelisTests {

    private static int iterations;
    private static final List<Device> devices = new ArrayList<>();
    private static final List<Speaker> speakers = new ArrayList<>();
    private static Config config = new BaseConfig();
    private static List<ProtelisNode> nodes;


    @BeforeAll
    static void init() {
        config.addSpec(ProtelisConfigSpec.spec);
        config = config.from().toml.resource("config.toml");
        String protelisModuleName = config.get(ProtelisConfigSpec.protelisModuleName);
        iterations = config.get(ProtelisConfigSpec.iterations);
        nodes = config.get(ProtelisConfigSpec.nodes);
        nodes.forEach(n -> {
            SocketNetworkManager netmgr = new SocketNetworkManager(new IntDeviceUID(n.getId()), n.getHostandport().getPort(), n.getNeighbors());
            try {
                netmgr.listen();
            } catch (IOException e) {
                e.printStackTrace();
            }
            ProtelisProgram program = ProtelisLoader.parse(protelisModuleName);
            Speaker speaker = Mockito.spy(new ConsoleSpeaker());
            speakers.add(speaker);
            Device node = new Device(program, n.getId(), netmgr, speaker);
            if (n.isLeader()) {
                node.getDeviceCapabilities().getExecutionEnvironment().put("leader", true);
            }
            devices.add(node);
        });
    }

    @Test
    void testExecution() {
        List<Integer> leaders = config.get(ProtelisConfigSpec.nodes).stream()
                .filter(ProtelisNode::isLeader)
                .map(ProtelisNode::getId)
                .collect(Collectors.toList());
        for (int i = 0; i < iterations; i++) {
            devices.forEach(Device::runCycle);
        }

        List<String> messages = DoubleStream.iterate(3f, i -> i - 1)
                .limit(iterations)
                .mapToObj(x -> "The leader's count is: " + x)
                .collect(Collectors.toList());

        leaders.stream()
                .map(x -> Mockito.verify(speakers.get(x)))
                .flatMap(mock -> messages.stream().map(msg -> new Pair<>(mock, msg)))
                .forEach(x -> x.getFirst().announce(x.getSecond()));

        leaders.stream()
                .map(x -> new Pair<>(Mockito.verify(speakers.get(x), times(iterations)),
                        "The leader is at " + x))
                .forEach(x -> (x.getFirst()).announce(x.getSecond()));

        leaders.stream()
                .map(x -> Arrays.asList((x + nodes.size() - 1) % nodes.size(), (x + 1) % nodes.size()))
                .flatMap(Collection::stream)
                .map(x -> new Pair<>(Mockito.verify(speakers.get(x), atLeastOnce()),
                        "Hello from the leader to its neighbor at " + x))
                .forEach(x -> x.getFirst().announce(x.getSecond()));
    }
}

File: 04-kotlin-socket/src/test/kotlin/HelloProtelisTests.kt
import com.uchuhimo.konf.Config
import demo.*
import io.kotlintest.Spec
import io.kotlintest.specs.StringSpec
import io.mockk.spyk
import io.mockk.verify
import org.protelis.lang.ProtelisLoader
import kotlin.streams.toList

class HelloProtelisTests : StringSpec() {

    private var devices: List<Device> = emptyList()
    private var speakers: List<Speaker> = emptyList()
    private val config = Config { addSpec(ProtelisConfigSpec) }
            .from.toml.resource("config.toml")
    private val protelisModuleName = config[ProtelisConfigSpec.protelisModuleName]
    private val iterations = config[ProtelisConfigSpec.iterations]
    private val nodes = config[ProtelisConfigSpec.nodes]
    private val leaders = config[ProtelisConfigSpec.nodes].stream()
            .filter{ it.leader }
            .map { it.id }
            .toList()
    override fun beforeSpec(spec: Spec) {
        nodes.forEach {
            val socketNetworkManager = SocketNetworkManager(IntDeviceUID(it.id), it.hostandport.port, it.neighbors).apply { listen() }
            val program = ProtelisLoader.parse(protelisModuleName)
            val s = spyk(ConsoleSpeaker())
            val d = Device(program, it.id, socketNetworkManager, s)
            if (it.leader) {
                d.deviceCapabilities.executionEnvironment.put("leader", true)
            }
            devices += d
            speakers += s
        }
        repeat(iterations) {
            devices.forEach { it.runCycle() }
        }
        devices.forEach { (it.netmgr as SocketNetworkManager).stop() }
    }

    init {
        "The leader count should be correct" {
            val messages = generateSequence(3f) { it - 1 }
                    .take(iterations)
                    .map { "The leader's count is: ${it}"}
                    .toList()
            leaders.stream()
                    .flatMap({ x -> messages.stream().map({ msg -> Pair<Int, String>(x, msg) }) })
                    .forEach { verify(exactly = 1) { speakers[it.first].announce((it.second)) } }
        }

        "The leaders should announce their id" {
            leaders.forEach { verify(exactly = iterations) { speakers[it].announce("The leader is at ${it}") } }
        }

        "The leader neighbors should say something" {
            leaders
                    .map { it }
                    .flatMap { listOf(
                            (it + nodes.size -1) % nodes.size,
                            (it + 1) % nodes.size) }
                    .distinct()
                    .forEach { verify(atLeast = 1) { speakers[it].announce("Hello from the leader to its neighbor at ${it}") } }
        }
    }
}